tilequeue/process.py

```diff
@@ -556,16 +562,56 @@ def format_coord(
 # the output.
 def process_coord(coord, nominal_zoom, feature_layers, post_process_data,
                   formats, unpadded_bounds, cut_coords, buffer_cfg,
-                  output_calc_spec, scale=4096, log_fn=None, max_zoom_with_changes=16):
+                  output_calc_spec, scale=4096, log_fn=None,
+                  max_zoom_with_changes=16):
     processed_feature_layers, extra_data = process_coord_no_format(
         feature_layers, nominal_zoom, unpadded_bounds, post_process_data,
         output_calc_spec, log_fn=log_fn)
 
     all_formatted_tiles, extra_data = format_coord(
-        coord, nominal_zoom, max_zoom_with_changes, processed_feature_layers, formats,
+        coord, nominal_zoom, max_zoom_with_changes,
+        processed_feature_layers, formats,
         unpadded_bounds, cut_coords, buffer_cfg, extra_data, scale)
 
-    return all_formatted_tiles, extra_data
+    # Because cut coord zoom 15 and 16 shares a common nominal_zoom 16,
+    # the special logic below is a necessary hack to make the
+    # current highest cut_coord zoom 16 follow the end_zoom:17 config in
+    # the post_process of queries.yaml in vector-datasource.
+    # If our highest supported zoom is not 16, the system might be
+    # broken, so we assert 16 here. Plus, we hardcoded 16 elsewhere too:
+    # https://github.com/tilezen/tilequeue/blob/43a4d4d1b101a4410660c23f1d41222e85aaa3ba/tilequeue/process.py#L382
+    assert max_zoom_with_changes == 16
```

Comment:
if this works, I think we can actually do something more like
```
if max_zoom_with_changes == nominal_zoom: 
```

and then base everything off of that.  Like right now this breaks for anyone with an alternate config

Reply:
In fact I intentionally use the hardcode value because of the comment I added above this line. 

In short, we don't know whether it shall work when max_zoom_with_changes is not 16, when it is not 16, it may not throw exception throughout the whole process, but maybe the result is logically wrong; and if this happens it is way harder to debug what's wrong since the code doesn't error out and the bug is hidden. 

Thus, we want to error out here by doing an assertion and postponing the generalization to future development. When people do want to let this handling work for non-16 value in the future, they will encounter this error here and they know what they are doing and will test and take of the result's correctness by correcting the assertion then.

Reply:
In other words, I think we don't need to spend time testing the result of max_zoom_with_changes != 16 now in this PR.

Reply:
I'm fine keeping the hard coded 16 here, as we hard code 16 a few other places and its easier to read.

But the below section will likely significantly increase the build time, because it's doing a repeat pass in cutting the tiles. Let's see the actual impact in a build.